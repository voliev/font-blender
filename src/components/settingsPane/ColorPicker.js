import React, { Component } from "react";
import classNames from "classnames";
import OutsideClickHandler from "react-outside-click-handler";
import { SketchPicker } from "react-color";
import PropTypes from "prop-types";
import { MIN_GAP } from "../../constants";
import { isPropTypeHexColor } from "../../utils";
import "./styles/ColorPicker.css";

class ColorPicker extends Component {
  constructor() {
    super();

    this.containerRef = React.createRef();

    this.state = {
      displayColorPicker: false
    };

    this.handleClick = this.handleClick.bind(this);
    this.handleOutsideClick = this.handleOutsideClick.bind(this);
  }

  handleClick() {
    this.setState(({ displayColorPicker }) => ({
      displayColorPicker: !displayColorPicker
    }));
  }

  handleOutsideClick() {
    this.setState({
      displayColorPicker: false
    });
  }

  getPickerPositionClass() {
    if (!this.containerRef.current) {
      return;
    }

    const viewportHeight = document.documentElement.clientHeight;
    const containerCoords = this.containerRef.current.getBoundingClientRect();
    const gap = viewportHeight - containerCoords.bottom;

    return gap > MIN_GAP ? "below" : "above";
  }

  getPicker() {
    if (!this.state.displayColorPicker) return null;

    const pickerPositionClass = this.getPickerPositionClass();
    const pickerClasses = classNames({
      picker: true,
      above: pickerPositionClass === "above",
      below: pickerPositionClass === "below"
    });

    const { color, colorChangeHandler } = this.props;
    return (
      <div className={pickerClasses}>
        <SketchPicker
          disableAlpha={true}
          color={color}
          onChangeComplete={colorChangeHandler}
        />
      </div>
    );
  }

  render() {
    const { color, disabled } = this.props;

    const componentClasses = classNames({
      ColorPicker: true,
      active: this.state.displayColorPicker,
      disabled
    });

    const colorPicker = this.getPicker();

    return (
      <OutsideClickHandler
        onOutsideClick={this.handleOutsideClick}
        disabled={!this.state.displayColorPicker}
      >
        <div className={componentClasses}>
          <div
            className="active-container"
            onClick={!disabled ? this.handleClick : null}
            ref={this.containerRef}
          >
            <div className="icon">
              <div className="color" style={{ background: color }} />
            </div>
          </div>
          {colorPicker}
        </div>
      </OutsideClickHandler>
    );
  }
}

ColorPicker.propTypes = {
  disabled: PropTypes.bool.isRequired,
  colorChangeHandler: PropTypes.func.isRequired,
  color: isPropTypeHexColor
};

export default ColorPicker;
