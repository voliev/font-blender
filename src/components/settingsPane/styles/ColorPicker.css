.ColorPicker {
  position: relative;
}

.ColorPicker .icon {
  background-color: var(--white);
  border-radius: var(--border-radius-sm);
  box-shadow: var(--inactive-input-shadow);
  cursor: pointer;
  display: inline-block;
  height: 2rem;
  padding: var(--space-xxs);
  width: 100%;
}

.ColorPicker.active .icon {
  box-shadow: var(--active-input-shadow);
  ;
  outline: none;
}

.ColorPicker.disabled .icon {
  background-color: var(--off-white);
  cursor: default;
}

.ColorPicker .color {
  border-radius: var(--border-radius-sm);
  box-shadow: var(--inactive-input-shadow);
  ;
  height: 100%;
  width: 100%;
}

/* Disables inline style set by the module */
.ColorPicker.disabled .color {
  background-color: var(--light-grey) !important;
}

.ColorPicker.active .picker {
  position: absolute;
  z-index: 100;
}

.ColorPicker.active .picker.above {
  bottom: 33px;
}

.ColorPicker.active .picker.below {
  top: 33px;
}

/* Container extra height fix */
.ColorPicker .active-container {
  height: 32px;
  margin-bottom: 32px;
}
