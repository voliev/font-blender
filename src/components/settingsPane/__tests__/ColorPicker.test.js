import React from "react";
import { mount, shallow } from "enzyme";
import OutsideClickHandler from "react-outside-click-handler";
import ColorPicker from "../ColorPicker";
import { SketchPicker } from "react-color";

const setup = (
  renderFn = shallow,
  propsOverrides = {},
  container = undefined
) => {
  const props = {
    disabled: false,
    colorChangeHandler: jest.fn(),
    color: "#333",
    ...propsOverrides
  };

  return {
    ...props,
    wrapper: container
      ? renderFn(<ColorPicker {...props} />, { attachTo: container })
      : renderFn(<ColorPicker {...props} />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should receive corresponding props", () => {
  const { color, disabled, colorChangeHandler, wrapper } = setup(mount);

  expect(typeof wrapper.prop("disabled")).toBe("boolean");
  expect(wrapper.prop("disabled")).toBe(disabled);

  expect(typeof wrapper.prop("colorChangeHandler")).toBe("function");
  expect(wrapper.prop("colorChangeHandler")).toBe(colorChangeHandler);

  expect(typeof wrapper.prop("color")).toBe("string");
  expect(wrapper.prop("color")).toBe(color);
});

it("should set corresponding component classes", () => {
  const { wrapper } = setup();

  expect(wrapper.find(".ColorPicker").hasClass("active")).toBeFalsy();
  expect(wrapper.find(".ColorPicker").hasClass("disabled")).toBeFalsy();

  wrapper.setState({ displayColorPicker: true });

  expect(wrapper.find(".ColorPicker").hasClass("active")).toBeTruthy();

  wrapper.setState({ displayColorPicker: false });

  expect(wrapper.find(".ColorPicker").hasClass("active")).toBeFalsy();

  wrapper.setProps({ disabled: true });

  expect(wrapper.find(".ColorPicker").hasClass("disabled")).toBeTruthy();
});

it("should set corresponding state on initial render", () => {
  const { wrapper } = setup();

  expect(wrapper.state("displayColorPicker")).toBeFalsy();
});

it("should handle color change", () => {
  const { colorChangeHandler, wrapper } = setup();

  wrapper.setState({ displayColorPicker: true });
  wrapper.find(SketchPicker).prop("onChangeComplete")();

  expect(colorChangeHandler).toHaveBeenCalled();
  // Note: No need to check handler arguments because they are
  //       controlled by `react-color` npm module
});

it("should set ref", () => {
  const { wrapper } = setup(mount);

  const ref = wrapper.instance().containerRef.current;

  const container = wrapper.find(".active-container").instance();

  expect(ref).not.toBeNull();
  expect(ref).toEqual(container);
});

it("should handle click outside of the color picker", () => {
  const handleOutsideClick = jest.spyOn(
    ColorPicker.prototype,
    "handleOutsideClick"
  );
  const { wrapper } = setup();

  wrapper.setState({ displayColorPicker: true });

  wrapper.find(OutsideClickHandler).prop("onOutsideClick")();

  expect(handleOutsideClick).toHaveBeenCalledTimes(1);
  expect(wrapper.state("displayColorPicker")).toBeFalsy();
});

it("should disable OutsideClickHandler if color picker is inactive", () => {
  const { wrapper } = setup();

  expect(wrapper.state("displayColorPicker")).toBeFalsy();
  expect(wrapper.find(OutsideClickHandler).prop("disabled")).toBeTruthy();
});

it("should render picker depending on the components` state value", () => {
  const { wrapper } = setup();

  expect(wrapper.find(SketchPicker).length).toBe(0);

  wrapper.setState({ displayColorPicker: true });

  expect(wrapper.find(SketchPicker).length).toBe(1);
});

it("should set picker visibility handler depending on `disabled` prop value", () => {
  const { wrapper } = setup();

  expect(typeof wrapper.find(".active-container").prop("onClick")).toBe(
    "function"
  );

  wrapper.setProps({ disabled: true });

  expect(wrapper.find(".active-container").prop("onClick")).toBeNull();
});

it("should handle click on icon container", () => {
  const handleClick = jest.spyOn(ColorPicker.prototype, "handleClick");
  const { wrapper } = setup();

  wrapper.find(".active-container").simulate("click");

  expect(handleClick).toHaveBeenCalledTimes(1);
});

it("should return corresponding string depending on the color picker icon positon", () => {
  const getPickerPositionClass = jest.spyOn(
    ColorPicker.prototype,
    "getPickerPositionClass"
  );
  const jsdom = require("jsdom");
  const { JSDOM } = jsdom;
  const dom = new JSDOM(`
    <!DOCTYPE html>
    <html>
      <body>
        <div class="target"></div>
      </body>
    </html>
  `);

  Object.defineProperty(document, "documentElement", {
    value: dom.window.document.documentElement
  });

  Object.defineProperty(dom.window.HTMLElement.prototype, "clientHeight", {
    value: 800
  });

  const container = dom.window.document.createElement("div");
  container.setAttribute("class", "container");

  const { wrapper } = setup(mount, undefined, container);

  const target = dom.window.document.querySelector(".target");
  target.appendChild(container);

  let bottom = 250;
  const containerRefContext = {
    containerRef: {
      current: {
        getBoundingClientRect() {
          return { bottom };
        }
      }
    }
  };

  wrapper.instance().getPickerPositionClass.call(containerRefContext);

  expect(getPickerPositionClass).toHaveLastReturnedWith("below");

  bottom = 550;

  wrapper.instance().getPickerPositionClass.call(containerRefContext);

  expect(getPickerPositionClass).toHaveLastReturnedWith("above");
});
