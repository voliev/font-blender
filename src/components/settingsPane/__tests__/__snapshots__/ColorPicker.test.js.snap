// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<OutsideClickHandler
  disabled={true}
  display="block"
  onOutsideClick={[Function]}
  useCapture={true}
>
  <div
    className="ColorPicker"
  >
    <div
      className="active-container"
      onClick={[Function]}
    >
      <div
        className="icon"
      >
        <div
          className="color"
          style={
            Object {
              "background": "#333",
            }
          }
        />
      </div>
    </div>
  </div>
</OutsideClickHandler>
`;
